### DISCUSSION TITLE
Neural backflow states for the Bose-Hubbard ground state

### DISCUSSION THREAD

**Discussion by chroli1** [netket/netket#2195] [General]
Hi,

I have been playing with the code accompanying [this paper](https://quantum-journal.org/papers/q-2025-06-17-1772/) for the ground state of the 2D square-lattice Bose-Hubbard model with fixed particle number. I have been struggling to get the parameters right for large systems (say, an 8x8 lattice).

In particular, I find that the energy and gradients for my initial wavefunction blow up, and the optimiser fails. The issue seems to be with the Monte Carlo sampler: if the Markov chains are poorly thermalised, states with vanishing amplitude are sampled, and the 1/\psi factor used when calculating expectation values and gradients diverges.

Is there any general advice/references on setting the MCMC parameters correctly? I think the main ones are:

- The total number of samples
- The number of parallel chains
- The sweep size (i.e. how often I record a sample as I move along the chain)
- The burn-in time (i.e. how many samples to discard at the start of the chain

**Comment by PhilipVinc** [comment]
cc @Z-Denis (who is possibly on leave AFAIK)

**Comment by Z-Denis** [comment]
Hi @chroli1 ! I will need a bit more information to be able to pinpoint the origin of the issue. For what nonlinearity is this happening? The stability of the approach will crucially depend on the regime. In the superfluid regime (U/t << 18), you should not encounter any issue with Monte Carlo and your chains should thermalise pretty fast with very low tau corr. Close to the phase transition, the situation is much trickier since the state is roughly a Kronecker delta around the Mott state. This means that, unless you do some fancy importance sampling, your acceptance will dramatically drop and, depending on the number of samples, you might even encounter an ill-defined QGT, causing your preconditioned forces to blow up. Finally, in the Mott regime, standard VMC is expected to fail.

I can list a few of the tricks I used in the paper:
- I first train a bare 2-body Jastrow. Then I create the Jastrow-backflow state, initialise the Jastrow weights to the learnt ones and initialise the weights of the last linear projection (a_\mu in the paper IIRC) to something really perturbative (say a_\mu ~N(\mu=0, \sigma = 1e-4)). This reduces the initial variance quite a lot when starting to train the backflow.
- I run the VMC on a GPU with 1024 chains. Having as many chains as possible is crucial here, especially close to the critical point, to mitigate intra-chain correlation times. Since I have only a few samples per chain, I probably bumped a bit the number of discarded samples per VMC step.
- I always keep an eye on the tau corr (make sure you set: `netket.config.netket_experimental_fft_autocorrelation = True` at the top of your script). You should increase the sweep size as you become less and less superfluid.
- Make sure you use the right Metropolis rule. If you work at fixed number of particles, you should use the one I provide in the accompanying repo. If you code it yourself, be really really careful with the log corrections of the rule.
- Don't forget to thermalise your state, also after adding the backflow; if you initialise wrongly your weights this thermalisation might take ages.

This is all I can think of, but I am happy to help you if you provide more information. I can send you a job file, if you want.

**Comment by Z-Denis** [comment]
Forgot to mention it, but the optimisation curves in these Bose-Hubbard experiments (with residual connections) are probably the cleanest I have ever obtained with VMC so the slightest instability should be a strong signal that something is wrong with the hyperparameters. Also forgot to ask: how many chains, samples per chain? what is your sweep size and number of discarded samples?

**Comment by chroli1** [comment]
Thanks both!

A bit more detail on the physics parameters of the model:

- I'm using U = t = 1, so deep in the superfluid regime, and no chemical potential, instead fixing the particle number at the level of the Fock space
- I use unit filling, so number of particles = number of sites
- I have tried both open and periodic BCs, seems to make no difference
- I used square lattices, looking at sizes from 2x2 up to 8x8

And on the methodology:

- Generally I kept things simple, so I just used the vmc.py script provided with your paper, didn't change anything except the parameters in the JSON file and commenting out the first part with the bare Jastrow ansatz (see below, I think this could be the issue actually!)
- In particular, I've not changed the approach to choosing the NN initial parameters or the Metropolis rule
- I'm running on a single GPU
- To narrow down the issue, I wrote some fresh code where I just generate some samples from your random initial state and evaluate the expectation value of the Hamiltonian, so no optimisation or much going on. From this, I worked out that the issue seems to be states with vanishing amplitude being sampled, leading to expectation values blowing up
- I have used a wide range of parameters and combinations to try to get things working. Long story short, even using a lot of samples (8192) and varying the number of chains, burn-in size and downsampling over wide ranges, I still find this issue (for context, I varied the number of chains from 1 up to all 8192 samples, and the burn-in and downsampling over orders of magnitude too)
- The only thing there that seems to help is making the downsampling really large (maybe)
- Even for parameters that produce 'good' energy EVs, sometimes the R-hat and \tau statistics have quite poor values

Your suggestions are really helpful. I was aware of some of this, which is reassuring! I think I am not doing this two-stage approach of training a bare Jastrow ansatz first, which might be the issue actually. I will try this out. 

Regardless, I think this is a more fundamental methodology issue and not just parameters - as you say, it shouldn't be that difficult.

**Comment by Z-Denis** [comment]
How do you initialise the chains? It is possible that the default `hilbert.random.random_state` yields configurations with extremely unbalanced populations that make the sampling of the energy and its forces very inaccurate with MC (large variance). The case of a warm-start from two-body Jastrow might be different should the bare two-body Jastrow be less sensitive to that, thus allowing a thermalisation to a decent configuration support before performing the whole optimisation procedure with the backflow transformation.

**Comment by Z-Denis** [comment]
Actually, having a look at the new constrained logic in NetKet, it seems to initialise the chains to the right thing (essentially the ideal superfluid):
https://github.com/netket/netket/blob/a8675bdc6f457b922262b0acfc5916183fc6c2c1/netket/hilbert/random/fock.py#L77-L109

It is very similar to the logic I used for the paper, if not identical. Another aspect that one might check is that the Jastrow weight matrix be positive and invertible, otherwise thermalising the chains will lead to exploding occupation variances (<n_i n_j> ~ [1/W]_{ij})

**Comment by chroli1** [comment]
This is now working! 

The issue is that I misunderstood the vmc.py script: I thought the plain Jastrow calculation was independent of the other one for comparison purposes, and commented that out to simplify things. I therefore wasn't doing the plain Jastrow pre-training, and unsurprisingly the optimisation of just the NN backflow part failed. Pretty basic error on my part but there we are!

One other general question: we are interested in looking into disorder in the Bose-Hubbard model. Do you expect that major changes will be needed to my code to do this? Perhaps changes to the sampler are needed?

Thanks so much for your help!

**Comment by Z-Denis** [comment]
That's great news!

Your project sounds interesting. I do think there are major changes to be considered. I depends a bit on the path you follow, but the model has been designed as to be translationally invariant. If you actually perform VMC for each realisation of the disorder, you will have to (at least) promote the Jastrow factor to a full dense symmetric matrix. 

Let me know if you want some data or any further help!

**Comment by chroli1** [comment]
Thanks! This makes sense - I suppose the translational invariance is baked in to the ResNet/convolutional neural network structure. Perhaps swapping to a different neural network entirely is needed?

**Comment by Z-Denis** [comment]
Before touching the backflow transformation, I would start breaking the translational invariance at the level of the Jastrow factor and by adding a spatial dependence to the final projection vector of the equivariant backflow features (`a_\mu` into `a_{i\mu}`).

**Comment by chroli1** [comment]
Thanks! This could be an interesting project in itself. If anything interesting comes of it I'll let you know. Thanks again for the help.

**Comment by Z-Denis** [comment]
To add to my previous comment: The CNN just makes the output equivariant, which is not necessarily bad in a disordered system. If you want to change the architecture of the backflow altogether you might try some vision transformer or whatever stable and scalable attention-based model you fancy.
